### PR TITLE
Adds proof harnesses for aws_byte_cursor_read* functions

### PR DESCRIPTION
**Pull request by feliperodri**
*Description of changes:*
1. Updates implementation of `aws_byte_cursor_read*` functions with pre- and post-conditions;
2. Adds proof harnesses for `aws_byte_cursor_read*` functions;

This PR depends on https://github.com/awslabs/aws-c-common/pull/385

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by feliperodri** [comment]
@tautschnig @karkhaz @angelhof @danielsn @markrtuttle could you review the proof harnesses?

File: .cbmc-batch/jobs/aws_byte_cursor_read_and_fill_buffer/aws_byte_cursor_read_and_fill_buffer_harness.c [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    __CPROVER_assume(aws_byte_cursor_is_valid(&cur));
+    ensure_byte_buf_has_allocated_buffer_member(&buf);
+    __CPROVER_assume(aws_byte_buf_is_valid(&buf));
+
+    /* save current state of the data structure */
+    struct aws_byte_cursor old_cur = cur;
+    struct store_byte_from_buffer old_byte_from_cur;
+    save_byte_from_array(cur.ptr, cur.len, &old_byte_from_cur);
+
+    /* save current state of the data structure */
```
Comment (tautschnig) [inline .cbmc-batch/jobs/aws_byte_cursor_read_and_fill_buffer/aws_byte_cursor_read_and_fill_buffer_harness.c]: I found having the same comment twice confusing - maybe just omit that second comment?

**Comment by feliperodri** [comment]
@JonathanHenson @ColdenCullen @justinboswell @bretambrose could you review the changes in `byte_buf.h`?

File: include/aws/common/byte_buf.h [hunk 2]
```diff
@@ -684,15 +684,20 @@ AWS_STATIC_IMPL struct aws_byte_cursor aws_byte_cursor_advance_nospec(
  * cursor unchanged.
  */
 AWS_STATIC_IMPL bool aws_byte_cursor_read(
-    struct aws_byte_cursor *AWS_RESTRICT cur,
-    void *AWS_RESTRICT dest,
+    struct aws_byte_cursor *const AWS_RESTRICT cur,
```
Comment (bretambrose) [inline include/aws/common/byte_buf.h]: Is this change necessary?  I did an informal, non-binding poll of some of the other devs here and no one really cares for this particular aspect of const (the address itself as opposed to the contents at the address) since it confuses the kind of const people normally care about and people tend to leave parameters alone anyways.  Is this useful for the prover?

Comment (feliperodri) [reply include/aws/common/byte_buf.h]: This is not required for any proofs, I just added to make it clear this function doesn't update the pointer. As the devs agreed this aspect it's not really relevant, I'll remove from this PR and I'll keep it in mind for the next ones.

**Comment by feliperodri** [comment]
@bretambrose I implemented your comments. Could you review the changes in `byte_buf.h` again?

**Comment by tautschnig** [comment]
@feliperodri Is this blocked on some bugs or is further work on this PR needed?

**Comment by feliperodri** [comment]
> @feliperodri Is this blocked on some bugs or is further work on this PR needed?

This is blocked on a CBMC issue.
